README.md
<!-- deno-fmt-ignore-file -->

Bencodex.js
===========

[![deno.land/x/bencodex status][]][deno.land/x/bencodex]
[![npm status][]][@planetarium/bencodex]
[![CI status][]][GitHub Actions]
[![Coverage status][]][Coveralls]

This library is an alternative take on implementing [Bencodex] in JavaScript.

It focused to address the following problems from the existing JavaScript
implementation(s):

 -  *No Node.js-specific APIs*:  It does not depend on Node.js-specific APIs
    such as [`Buffer`].  Usable with web browsers, Deno, and Node.js.

 -  *Static-time and runtime type safety*:  It is fully written in TypeScript,
    and promise you, we use [`any`] nowhere, and it never trusts any data from
    external sources thoughtlessly.  Instead, it tries to parse data and
    validate them whenever it reads data from external sources.  It's not only
    written in TypeScript, it also has runtime checks here and there so that
    you could use it in vanilla JavaScript without fear.

 -  *Well-tested*:  Passing only the Bencodex spec test suite is not enough to
    guarantee the whole functionality of a library, as every library has their
    own part to interpret data in JavaScript (how data should look like in
    JavaScript values, where data are read from and written into, whether data
    are streamed or given in a single big chunk, etc).  It has its own unit
    tests besides spec test suite, and we've done the best to reach
    [near-100% test coverage][Coveralls].

 -  *Comprehensive docs*:  Every non-trivial library needs the three types of
    docs: tutorials for getting stated, in-depth manuals for common scenarios,
    and complete API references with concise examples.  It has all of them!

 -  *Reducing unnecessary memcpy*:  Instead of creating chunks of
    [`Uint8Array`]s and then allocating a large buffer to concatenate the whole
    data once again, it allocates only the necessary buffer only once.
    It is not just memory-efficient, also time-efficient as memcpy quite
    takes time.

 -  *Proper binary keys*:  It does not simply depend on [`Map`], which compares
    its [`Uint8Array`] keys by reference equality.  Instead, it implements its
    own proper dictionary which compares [`Uint8Array`] by content equality.

Distributed under LGPL 2.1 or later.

[deno.land/x/bencodex status]: https://img.shields.io/github/v/tag/planetarium/bencodex.js?label=deno.land%2Fx%2Fbencodex
[deno.land/x/bencodex]: https://deno.land/x/bencodex
[npm status]: https://img.shields.io/npm/v/@planetarium/bencodex?label=npm%20i%20%40planetarium%2Fbencodex
[@planetarium/bencodex]: https://www.npmjs.com/package/@planetarium/bencodex
[CI status]: https://github.com/planetarium/bencodex.js/actions/workflows/main.yaml/badge.svg?branch=main
[GitHub Actions]: https://github.com/planetarium/bencodex.js/actions/workflows/main.yaml
[Coverage status]: https://coveralls.io/repos/github/planetarium/bencodex.js/badge.svg
[Coveralls]: https://coveralls.io/github/planetarium/bencodex.js
[Bencodex]: https://bencodex.org/
[`Buffer`]: https://nodejs.org/api/buffer.html
[`any`]: https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#any
[`Uint8Array`]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array
[`Map`]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Map


Getting started
---------------

See also the [complete API references] as well.

[complete API references]: https://deno.land/x/bencodex/mod.ts


### Deno

It's available on [deno.land/x/bencodex]:

~~~~ typescript
import { decode, encode } from "https://deno.land/x/bencodex/mod.ts";

const value = new Map([
  ["foo", true],
  ["bar", null],
  ["baz", new Uint8Array([0x73, 0x70, 0x61, 0x6d])],
  [
    new Uint8Array([0x68, 0x65, 0x6c, 0x6c, 0x6f]),
    [
      123456n,
      "Unicode string",
      false,
    ],
  ],
]);
const encoded = encode(value);
const decoded = decode(encoded);
~~~~


### Node.js

Add [@planetarium/bencodex] to your dependencies using your favorite package
manager:

~~~~ console
npm install @planetarium/bencodex
~~~~

The API usage is equivalent to the example above for Deno, except that you
need to import things from `"@planetarium/bencodex"` instead of the
*deno.land/x/* URL:

~~~~ typescript
import { decode, encode } from "@planetarium/bencodex";
~~~~


Types
-----

| Bencodex   | TypeScript ([`Value`])        |
|------------|-------------------------------|
| Null       | `null`                        |
| Boolean    | `boolean`                     |
| Integer    | `bigint` (not `number`)       |
| Binary     | [`Uint8Array`]                |
| Text       | `string`                      |
| List       | `Value[]`                     |
| Dictionary | [`Dictionary`]                |

[`Value`]: https://deno.land/x/bencodex/mod.ts?s=Value
[`Dictionary`]: https://deno.land/x/bencodex/mod.ts?s=Dictionary


Benchmarks
----------

~~~~ console
$ deno bench
cpu: 12th Gen Intel(R) Core(TM) i5-1235U
runtime: deno 1.30.3 (x86_64-unknown-linux-gnu)

file:///home/dahlia/src/planetarium/bencodex.js/bench/comparison.bench.ts
benchmark                        time (avg)             (min … max)       p75       p99      p995
------------------------------------------------------------------- -----------------------------
encoding (bencodex.js)        28.96 µs/iter   (17.24 µs … 10.43 ms)  22.25 µs  117.5 µs 221.98 µs
encoding (disjukr/bencodex)   66.01 µs/iter    (39.36 µs … 4.38 ms)  55.58 µs  215.2 µs 313.14 µs

summary
  encoding (bencodex.js)
   2.28x faster than encoding (disjukr/bencodex)

decoding (bencodex.js)         7.73 µs/iter    (6.35 µs … 16.97 ms)   6.92 µs  17.39 µs  30.93 µs
decoding (disjukr/bencodex)   45.76 µs/iter     (31.2 µs … 5.08 ms)  46.35 µs 135.29 µs 185.85 µs

summary
  decoding (bencodex.js)
   5.92x faster than decoding (disjukr/bencodex)
~~~~

You can run the benchmarks by `deno bench`.


CHANGES.md
<!-- deno-fmt-ignore-file -->

Changelog
=========

Version 0.3.0
-------------

To be released.

 -  Added `SizeEstimationOptions` interface.


Version 0.2.2
-------------

Released on March 6, 2023.  This is a hotfix of 0.2.1.

 -  Fixed a bug that [`areValuesEqual()`] and [`areDictionariesEqual()`] were
    not re-exported from *mod.ts* and npm [@planetarium/bencodex] package.

[`areValuesEqual()`]:
https://deno.land/x/bencodex@0.2.2/mod.ts?s=areValuesEqual
[`areDictionariesEqual()`]:
https://deno.land/x/bencodex@0.2.2/mod.ts?s=areDictionariesEqual
[@planetarium/bencodex]: https://www.npmjs.com/package/@planetarium/bencodex


Version 0.2.1
-------------

Released on March 6, 2023.  This release was mistaken.  Don't use this version,
and use 0.2.2 instead.

 -  <del>Fixed a bug that [`areValuesEqual()`] and [`areDictionariesEqual()`]
    were not re-exported from *mod.ts* and npm [@planetarium/bencodex]
    package.</del>


Version 0.2.0
-------------

Released on March 3, 2023.

 -  Added [`areValuesEqual()`][] function.
 -  Added [`areDictionariesEqual()`][] function.


Version 0.1.1
-------------

Released on March 3, 2023.

 -  [`BencodexDictionary`] and [`RecordView`] now shows easier inspection
    representions on the eyes both on Deno ([`Deno.inspect()`]) and
    Node.js ([`util.inspect()`]).


[`BencodexDictionary`]: https://deno.land/x/bencodex@0.1.1/mod.ts?s=BencodexDictionary
[`RecordView`]: https://deno.land/x/bencodex@0.1.1/mod.ts?s=RecordView
[`Deno.inspect()`]: https://deno.land/api?s=Deno.inspect
[`util.inspect()`]: https://nodejs.org/api/util.html#utilinspectobject-options


Version 0.1.0
-------------

Initial release. Released on March 2, 2023.


src/types.ts
/**
 * This module provides a basic type definitions for Bencodex, and some utility
 * functions for them.
 *
 * @module
 */

import { BencodexDictionary } from "./dict.ts";
import { areUint8ArraysEqual, compareUint8Arrays } from "./utils.ts";

/**
 * Represents a value which can be used as a key in a Bencodex dictionary.
 * It is either a string ("text" in Bencodex) or a `Unit8Array` instance
 * ("binary" in Bencodex).
 *
 * Note that every {@link Key} is also a {@link Value} (not vice versa).
 *
 * Here are examples of valid {@link Key} values:
 *
 * ```typescript
 * let text: Key = "key";
 * let binary: Key = new Uint8Array([0x6b, 0x65, 0x79]);
 * ```
 */
export type Key = string | Uint8Array;

/**
 * Represents a value which can be encoded in Bencodex.  Each data type in
 * Bencodex has its corresponding type in JavaScript:
 *
 * - `null` represents Bencodex's null
 * - `boolean` represents Bencodex's Boolean
 * - `bigint` represents Bencodex's integer (note that `number` is not a valid
 *   type for Bencodex's integer)
 * - `string` represents Bencodex's text
 * - `Uint8Array` represents Bencodex's binary
 * - `Value[]` represents Bencodex's list
 * - {@link Dictionary} represents Bencodex's dictionary (note that it is not
 *   a concrete type, but an interface)
 *
 * Here are examples of valid {@link Value}s:
 *
 * ```typescript
 * let null_: Value = null;
 * let bool: Value = true;
 * let integer: Value = 123n;  // note that it ends with suffix `n`
 * let text: Value = "value";
 * let binary: Value = new Uint8Array([0x76, 0x61, 0x6c, 0x75, 0x65]);
 * let list: Value = [
 *   null,
 *   true,
 *   123n,
 *   "value",
 *   new Uint8Array([0x76, 0x61, 0x6c, 0x75, 0x65]),
 *   [],
 *   new Map(),
 * ];
 * let dictionary: Value = new Map([
 *   ["string key", list],
 *   [new Uint8Array([0x6b, 0x65, 0x79]), "<- binary key"]
 * ]);
 * ```
 */
export type Value = null | boolean | bigint | Key | List | Dictionary;

/**
 * Represents a Bencodex list.  It basically is a read-only `Value[]`.
 */
export type List = readonly Value[];

/**
 * Represents a Bencodex dictionary.  It basically behaves like a read-only
 * `Map<Key, Value>`, but it is not necessarily a `Map` instance.
 *
 * @see {@link BencodexDictionary}
 * @see {@link RecordView}
 */
export interface Dictionary extends Iterable<readonly [Key, Value]> {
  /**
   * The number of key-value pairs in this dictionary.
   */
  readonly size: number;

  /**
   * Gets the value associated with the given key.
   * @param key The key to get the value associated with.
   * @returns The value associated with the given key, or `undefined` if the
   *         given key is not present in this dictionary.
   */
  get(key: Key): Value | undefined;

  /**
   * Checks if the given key is present in this dictionary.
   * @param key The key to check.
   * @returns `true` iff the given key is present in this dictionary.
   */
  has(key: Key): boolean;

  /**
   * Gets an iterable object which iterates over the keys in this dictionary.
   * @returns An iterable object which iterates over the keys in this
   *          dictionary.
   */
  keys(): Iterable<Key>;

  /**
   * Gets an iterable object which iterates over the values in this dictionary.
   * @returns An iterable object which iterates over the values in this
   *          dictionary.
   */
  values(): Iterable<Value>;

  /**
   * Gets an iterable object which iterates over the key-value pairs in this
   * dictionary.  This method is equivalent to `Dictionary[Symbol.iterator]()`.
   * @returns An iterable object which iterates over the key-value pairs in
   *          this dictionary.
   */
  entries(): Iterable<readonly [Key, Value]>;

  /**
   * Calls the given callback function for each key-value pair in this
   * dictionary.
   * @param callback The callback function to call for each key-value pair.
   *                 This function is called with three arguments: the value,
   *                 the key, and the dictionary itself.
   * @param thisArg The value to use as `this` when calling the callback
   *                function.
   */
  forEach(
    callback: (value: Value, key: Key, dictionary: Dictionary) => void,
    thisArg?: unknown,
  ): void;

  /**
   * Gets an iterator which iterates over the key-value pairs in this
   * dictionary.  This method is equivalent to `Dictionary.entries()`.
   * @returns An iterator which iterates over the key-value pairs in this
   *          dictionary.
   * @see {@link Dictionary.entries}
   */
  [Symbol.iterator](): Iterator<readonly [Key, Value]>;
}

/**
 * Checks if the given value is a Bencodex dictionary.
 *
 * In TypeScript, this function can be used as a type guard.
 * @param value The value to check.
 * @returns `true` iff the given value is a Bencodex dictionary.
 */
export function isDictionary(value: unknown): value is Dictionary {
  return (
    typeof value === "object" && value != null &&
    "size" in value && typeof value.size === "number" &&
    "get" in value && typeof value.get === "function" &&
    "has" in value && typeof value.has === "function" &&
    "keys" in value && typeof value.keys === "function" &&
    "values" in value && typeof value.values === "function" &&
    "entries" in value && typeof value.entries === "function" &&
    "forEach" in value && typeof value.forEach === "function" &&
    Symbol.iterator in value && typeof value[Symbol.iterator] === "function"
  );
}

/**
 * Checks if the given value can be used as a key in a Bencodex dictionary.
 *
 * In TypeScript, this function can be used as a type guard.
 * @param value The value to check.
 * @returns `true` iff the given value can be used as a key in a Bencodex
 *          dictionary.
 */
export function isKey(value: unknown): value is Key {
  return typeof value === "string" || value instanceof Uint8Array;
}

/**
 * Checks if the given keys have the same type and the same contents.  In other
 * words, this function checks if the given keys are encoded in the same
 * Bencodex data.
 * @param a A key to compare.
 * @param b Another key to compare.
 * @returns `true` iff the given keys have the same type and the same contents.
 *          It returns `false` when the given keys have invalid types even if
 *          they have the same contents, e.g., `areKeysEqual(1n, 1n)` returns
 *          `false`.
 */
export function areKeysEqual(a: Key, b: Key): boolean {
  if (typeof a === "string") {
    return typeof b === "string" && a === b;
  } else if (a instanceof Uint8Array) {
    return b instanceof Uint8Array && areUint8ArraysEqual(a, b);
  }
  return false;
}

/**
 * Compares two keys in the specified order in the Bencodex specification.
 *
 * This function can be passed to `Array.prototype.sort()` to sort keys in the
 * order specified in the Bencodex specification.
 * @param a A key to compare.
 * @param b Another key to compare.
 * @returns A negative number if `a` is former than `b`, zero if `a` is equal to
 *         `b`, or a positive number if `a` is latter than `b`.
 * @throws {TypeError} When any of the given keys is neither a `string` nor
 *         a `Uint8Array`.
 */
export function compareKeys(a: Key, b: Key): number {
  if (typeof a === "string") {
    if (typeof b === "string") return a < b ? -1 : a === b ? 0 : 1;
    else if (b instanceof Uint8Array) return 1;
    throw new TypeError(`Invalid key type: ${typeof b}`);
  } else if (a instanceof Uint8Array) {
    if (typeof b === "string") return -1;
    else if (b instanceof Uint8Array) return compareUint8Arrays(a, b);
    throw new TypeError(`Invalid key type: ${typeof b}`);
  }
  throw new TypeError(`Invalid key type: ${typeof a}`);
}

/**
 * Checks if the given Bencodex dictionaries have the same keys and the same
 * associated values.  In other words, this function checks if the given
 * dictionaries are encoded in the same Bencodex data.
 * @param a A dictionary to compare.
 * @param b Another dictionary to compare.
 * @returns `true` iff the given dictionaries have the same keys and the same
 *          associated values.  It returns `false` when the given dictionaries
 *          have invalid types even if they have the same keys and the same
 *          associated values, e.g., `areDictionariesEqual(1n, 1n)` returns
 *          `false`.
 */
export function areDictionariesEqual(a: Dictionary, b: Dictionary): boolean {
  if (!isDictionary(a) || !isDictionary(b)) return false;
  if (a.size !== b.size) return false;
  let otherEntries: undefined | (readonly [Key, Value])[];
  for (const [key, value] of a) {
    if (typeof key === "string") {
      const otherValue = b.get(key);
      if (typeof otherValue === "undefined") return false;
      if (!areValuesEqual(value, otherValue)) return false;
    } else {
      let otherValue = b.get(key);
      // As not every Dictionary implementation guarantees they compare binary
      // keys by their contents rather than by their references (e.g., Map),
      // we can't rely on their get() or has() methods for binary keys:
      if (typeof otherValue === "undefined") {
        if (otherEntries === undefined) otherEntries = [...b.entries()];
        for (const [bKey, bVal] of otherEntries) {
          if (areKeysEqual(key, bKey)) {
            otherValue = bVal;
            break;
          }
        }
      }
      if (typeof otherValue === "undefined") return false;
      if (!areValuesEqual(value, otherValue)) return false;
    }
  }
  return true;
}

/**
 * Checks if the given Bencodex values have the same type and the same contents.
 * In other words, this function checks if the given values are encoded in the
 * same Bencodex data.
 * @param a A Bencodex value to compare.
 * @param b Another Bencodex value to compare.
 * @returns `true` iff the given values have the same type and the same
 *          contents.  It returns `false` when the given values have invalid
 *          types even if they have the same contents, e.g., `areValuesEqual(1,
 *          1)` returns `false` (since `number`s are not a valid Bencodex type).
 */
export function areValuesEqual(a: Value, b: Value): boolean {
  if (a === null && b === null) return true;
  if (
    typeof a === "boolean" && typeof b === "boolean" ||
    typeof a === "bigint" && typeof b === "bigint"
  ) return a === b;
  if (isKey(a) && isKey(b)) return areKeysEqual(a, b);
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    return a.every((v, i) => areValuesEqual(v, b[i]));
  }
  // Note that areDictionariesEqual() checks the types of the given values
  // at runtime, and it returns false when the given values aren't dictionaries:
  return areDictionariesEqual(
    a as unknown as Dictionary,
    b as unknown as Dictionary,
  );
}
